iceoryx_posh/source/roudi/port_manager.cpp

```diff
@@ -225,14 +233,175 @@ void PortManager::doDiscoveryForSubscriberPort(SubscriberPortType& subscriberPor
+{
+    cxx::Ensures(serverPortData != nullptr && "serverPortData must not be a nullptr");
+
+    // create temporary server ports to orderly shut this server down
+    popo::ServerPortRouDi serverPortRoudi{*serverPortData};
+    popo::ServerPortUser serverPortUser{*serverPortData};
+
+    serverPortRoudi.releaseAllChunks();
+    serverPortUser.stopOffer();
+
+    // process STOP_OFFER for this server in RouDi and distribute it
+    serverPortRoudi.tryGetCaProMessage().and_then([this, &serverPortRoudi](auto caproMessage) {
+        cxx::Ensures(caproMessage.m_type == capro::CaproMessageType::STOP_OFFER);
+        cxx::Ensures(caproMessage.m_serviceType == capro::CaproServiceType::SERVER);
+
+        /// @todo iox-#27 report to port introspection
+        this->removeServerFromServiceRegistry(caproMessage.m_serviceDescription);
+        this->sendToAllMatchingClientPorts(caproMessage, serverPortRoudi);
+        this->sendToAllMatchingInterfacePorts(caproMessage);
+    });
+
+    /// @todo iox-#27 remove from port introspection
+
+    LogDebug() << "Destroy server port from runtime '" << serverPortData->m_runtimeName
+               << "' and with service description '" << serverPortData->m_serviceDescription << "'";
+
+    // delete server port from list after STOP_OFFER was processed
+    m_portPool->removeServerPort(serverPortData);
+}
+
+void PortManager::handleServerPorts() noexcept
+{
+    // get the changes of server port offer state
+    for (auto serverPortData : m_portPool->getServerPortDataList())
+    {
+        popo::ServerPortRouDi serverPort(*serverPortData);
+
+        doDiscoveryForServerPort(serverPort);
+
+        // check if we have to destroy this server port
+        if (serverPort.toBeDestroyed())
+        {
+            destroyServerPort(serverPortData);
+        }
+    }
+}
+
+void PortManager::doDiscoveryForServerPort(popo::ServerPortRouDi& serverPort) noexcept
+{
+    serverPort.tryGetCaProMessage().and_then([this, &serverPort](auto caproMessage) {
+        /// @todo iox-#27 report to port instrospection
+
+        if (capro::CaproMessageType::OFFER == caproMessage.m_type)
+        {
+            this->addServerToServiceRegistry(caproMessage.m_serviceDescription);
+        }
+        else if (capro::CaproMessageType::STOP_OFFER == caproMessage.m_type)
+        {
+            this->removeServerFromServiceRegistry(caproMessage.m_serviceDescription);
+        }
+        else
+        {
+            LogWarn() << "CaPro protocol error for server from runtime '" << serverPort.getRuntimeName()
+                      << "' and with service description '" << serverPort.getCaProServiceDescription()
+                      << "'! Cannot handle CaProMessageType '" << caproMessage.m_type << "'";
+            errorHandler(
+                Error::kPORT_MANAGER__HANDLE_SERVER_PORTS_INVALID_CAPRO_MESSAGE, nullptr, iox::ErrorLevel::MODERATE);
+        }
+
+        this->sendToAllMatchingClientPorts(caproMessage, serverPort);
```

Comment:
They also would get the messages that are a protocol error, but their state machine can handle

Reply:
In this case not, since the `ServerPort::tryGetCaProMessage` will either return a `nullopt`, `OFFER` or `STOP_OFFER`

Reply:
@elBoberido Which means that we'll never be in the else branch. But if `ServerPort::tryGetCaProMessage` would return something else, it is like I described, or? But fine for me as this is all checked on all sides

Reply:
@budrus yes. I copied it from the publisher port without thinking too much. The else branch should not be taken but it would nevertheless be a good idea to just return in the else branch.

Reply:
Fixed